hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -0,0 +1,244 @@
+              getTargetDatanodes(replicas, container, missingIndexes.size());
+
+          List<ReconstructECContainersCommand.DatanodeDetailsAndReplicaIndex>
+              sourceDatanodesWithIndex = new ArrayList<>();
+          for (ContainerReplica srcReplica : sources) {
+            sourceDatanodesWithIndex.add(
+                new ReconstructECContainersCommand
+                    .DatanodeDetailsAndReplicaIndex(
+                    srcReplica.getDatanodeDetails(),
+                    srcReplica.getReplicaIndex()));
+          }
+
+          final ReconstructECContainersCommand reconstructionCommand =
+              new ReconstructECContainersCommand(id.getProtobuf().getId(),
+                  sourceDatanodesWithIndex, selectedDatanodes,
+                  int2byte(missingIndexes),
+                  repConfig);
+          // Keeping the first target node as coordinator.
+          return ImmutableMap.of(selectedDatanodes.get(0),
+              reconstructionCommand);
+        } else {
+          LOG.warn("Cannot proceed for EC container reconstruction for {}, due"
+              + "to insufficient source replicas found. Number of source "
+              + "replicas needed: {}. Number of available source replicas are:"
+              + " {}. Available sources are: {}", container.containerID(),
+              repConfig.getData(), sources.size(), sources);
+        }
+      } else if (result instanceof ContainerHealthResult
+          .UnderReplicatedHealthResult && ((ContainerHealthResult
+          .UnderReplicatedHealthResult) result)
+          .underReplicatedDueToDecommission()) {
+        Map<DatanodeDetails, SCMCommand> commands = new HashMap<>();
+        Set<Integer> decomIndexes = replicaCount.decommissioningIndexes();
+        final List<DatanodeDetails> selectedDatanodes =
+            getTargetDatanodes(replicas, container, decomIndexes.size());
+        Iterator<DatanodeDetails> iterator = selectedDatanodes.iterator();
+        // In this case we need to do one to one copy.
+        for (ContainerReplica replica : replicas) {
+          if (decomIndexes.contains(replica.getReplicaIndex())) {
+            if (!iterator.hasNext()) {
+              LOG.warn("Couldn't find the enough targets. Available source"
+                  + " nodes: {}, the target nodes: {} and the decommission"
+                  + " indexes: {}", replicas, selectedDatanodes, decomIndexes);
+              break;
+            }
+            DatanodeDetails decommissioningSrcNode
+                = replica.getDatanodeDetails();
+            final ReplicateContainerCommand replicateCommand =
+                new ReplicateContainerCommand(id.getProtobuf().getId(),
+                    ImmutableList.of(decommissioningSrcNode));
+            DatanodeDetails target = iterator.next();
+            commands.put(target, replicateCommand);
+          }
+        }
+        return commands;
+      } else {
+        if (replicaCount.unRecoverable()) {
```

Comment:
Might be cleaner if we lifted these checks to the top of the method along with the others, as they they are all groups together.